include/deal.II/non_matching/mapping_info.h

```diff
@@ -646,288 +1081,362 @@ namespace NonMatching
+  template <int dim, int spacedim, typename Number>
+  inline const ArrayView<const Point<dim, Number>>
+  MappingInfo<dim, spacedim, Number>::get_real_points(
+    const unsigned int cell_index,
+    const unsigned int face_number) const
   {
-    update_flags_mapping |=
-      mapping->requires_update_flags(update_flags_mapping);
-
-    mapping_data.initialize(quadrature.size(), update_flags_mapping);
-
-    // reuse internal_mapping_data for MappingQ to avoid memory allocations
-    if (const MappingQ<dim, spacedim> *mapping_q =
-          dynamic_cast<const MappingQ<dim, spacedim> *>(&(*mapping)))
+    if (cell_index == numbers::invalid_unsigned_int &&
+        face_number == numbers::invalid_unsigned_int)
       {
-        (void)mapping_q;
-        auto &data =
-          dynamic_cast<typename MappingQ<dim, spacedim>::InternalData &>(
-            *internal_mapping_data);
-        data.initialize(update_flags_mapping, quadrature, quadrature.size());
+        Assert(state == State::single_cell,
+               ExcMessage(
+                 "This mapping info is not reinitialized for a single cell!"));
+        return make_array_view(real_points);
       }
     else
       {
-        internal_mapping_data =
-          mapping->get_data(update_flags_mapping, quadrature);
+        const unsigned int offset =
+          calculate_geometry_index_offset(cell_index, face_number);
+        const unsigned int data_index_offset = data_index_offsets[offset];
+        const unsigned int n_q_points =
+          data_index_offsets[offset + 1] - data_index_offset;
+        return make_array_view(real_points, data_index_offset, n_q_points);
       }
-
-    mapping->fill_fe_immersed_surface_values(cell,
-                                             quadrature,
-                                             *internal_mapping_data,
-                                             mapping_data);
   }
 
 
 
-  template <int dim, int spacedim>
-  void
-  MappingInfo<dim, spacedim>::compute_mapping_data_for_face_quadrature(
-    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
-    const unsigned int                                          face_no,
-    const Quadrature<dim - 1> &                                 quadrature,
-    MappingData &                                               mapping_data)
+  template <int dim, int spacedim, typename Number>
+  unsigned int
+  MappingInfo<dim, spacedim, Number>::calculate_unit_point_index_offset(
+    const unsigned int cell_index,
+    const unsigned int face_number) const
   {
-    update_flags_mapping |=
-      mapping->requires_update_flags(update_flags_mapping);
+    if (cell_index == numbers::invalid_unsigned_int &&
+        face_number == numbers::invalid_unsigned_int)
+      {
+        Assert(state == State::single_cell,
+               ExcMessage(
+                 "This mapping info is not reinitialized for a single cell!"));
+        return 0;
```

Comment:
Who stores the data in this case?

Reply:
Data is always stored in `MappingInfo` also for a single cell.